concrete/src/Entity/File/Version.php

```diff
@@ -657,697 +872,862 @@ public function getVersionLogComments()
+                        try {
+                            $imageLibrary->setMetadataReader(new ExifMetadataReader());
+                        } catch (NotSupportedException $e) {
+                        }
                     }
                 }
+                try {
+                    $this->imagineImage = $imageLibrary->load($resource->read());
+                } catch (FileNotFoundException $e) {
+                    $this->imagineImage = false;
+                }
             }
-
-            $this->imagineImage = $imageLibrary->load($resource->read());
         }
 
-        return $this->imagineImage;
+        return $this->imagineImage ?: null;
     }
 
     /**
-     * Unload the loaded image.
+     * Unload the loaded Image instance.
      */
     public function releaseImagineImage()
     {
         $this->imagineImage = null;
     }
 
     /**
-     * Rescan the thumbnails for a file (images only).
+     * Delete and re-create all the thumbnail types (only applicable to image files).
      *
-     * @return bool False on failure
+     * @return bool return true on success, false on failure (file is not an image, problems during image processing, ...)
      */
     public function rescanThumbnails()
     {
-        if ($this->fvType != \Concrete\Core\File\Type\Type::T_IMAGE) {
-            return false;
-        }
-
-        $imagewidth = $this->getAttribute('width');
-        $imageheight = $this->getAttribute('height');
-        $types = Type::getVersionList();
-
-        try {
-            $image = $this->getImagineImage();
+        $result = false;
+        if ($this->fvType == \Concrete\Core\File\Type\Type::T_IMAGE) {
+            try {
+                $image = $this->getImagineImage();
+                if ($image) {
+                    $imagewidth = $this->getAttribute('width') ?: $image->getSize()->getWidth();
+                    $imageheight = $this->getAttribute('height') ?: $image->getSize()->getHeight();
+                    $types = Type::getVersionList();
+                    foreach ($types as $type) {
+                        // delete the file if it exists
+                        $this->deleteThumbnail($type);
+
+                        // if image is smaller than size requested, don't create thumbnail
+                        if ($imagewidth < $type->getWidth() && $imageheight < $type->getHeight()) {
+                            continue;
+                        }
 
-            if ($image) {
-                /* @var \Imagine\Imagick\Image $image */
-                if (!$imagewidth) {
-                    $imagewidth = $image->getSize()->getWidth();
-                }
-                if (!$imageheight) {
-                    $imageheight = $image->getSize()->getHeight();
-                }
+                        // This should not happen as it is not allowed when creating thumbnail types and both width and heght
```

Comment:
when -> if, and 'height'

Reply:
IMHO `it is not allowed when creating thumbnail types` is correct (`it is not allowed if creating thumbnail types` means something different)